File: ReverseVowelsOfAString/reverse_vowels_of_a_string.go
package main

import "unicode"

// Time Complexity: O(N)
// Space Complexity: O(N)

// fun fact:
// `Y` and `y` can be a vowel as well sometimes.
// glad the problem statement defines it well
func isVowel(c rune) bool {
	// alternatively, we can just check
	// return c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U' ||
	//        c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
	c = unicode.ToLower(c)
	return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
}

func reverseVowels(s string) string {
	ss := []rune(s)
	n := len(ss)
	l, r := 0, n-1
	for l < r {
		// find the index of the first vowel in the given range
		for l < r && !isVowel(ss[l]) {
			l += 1
		}
		// find the index of last vowel in the given range
		for r > l && !isVowel(ss[r]) {
			r -= 1
		}
		// swap ss[l] and ss[r]
		ss[l], ss[r] = ss[r], ss[l]
		// since we've processed the character ss[l],
		// we move the left pointer to the right
		l += 1
		// since we've processed the character ss[r],
		// we move the right pointer to the left
		r -= 1
	}
	return string(ss)
}

File: ReverseVowelsOfAString/reverse_vowels_of_a_string.dart
/*


-* 345. Reverse Vowels of a String *-

Given a string s, reverse only all the vowels in the string and return it.

The vowels are 'a', 'e', 'i', 'o', and 'u', and they can appear in both lower and upper cases, more than once.



Example 1:

Input: s = "hello"
Output: "holle"
Example 2:

Input: s = "leetcode"
Output: "leotcede"


Constraints:

1 <= s.length <= 3 * 105
s consist of printable ASCII characters.
Accepted
403,544
Submissions
838,146


*/

class A {
  String reverseVowels(String s) {
    List<String> char = s.split("");
    int left = 0;
    int right = s.length - 1;
    while (left < right) {
      while (left < right && !isVowels(char[left])) {
        left++;
      }
      while (left < right && !isVowels(char[right])) {
        right--;
      }
      if (left < right) {
        String temp = char[left];
        char[left++] = char[right];
        char[right--] = temp;
      }
    }
    return char.join();
  }

  bool isVowels(String s) {
    return s == 'a' ||
        s == 'e' ||
        s == 'i' ||
        s == 'o' ||
        s == 'u' ||
        s == 'A' ||
        s == 'E' ||
        s == 'I' ||
        s == 'O' ||
        s == 'U';
  }
}

class B {
  // Set<String> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
  String vowels = 'aeiouAEIOU';
  String reverseVowels(String s) {
    List<String> char = s.split("");
    for (int i = 0, j = char.length - 1; i < j; i++, j--) {
      while (!vowels.contains(char[i]) && i < j) {
        i++;
      }

      while (!vowels.contains(char[j]) && i < j) {
        j--;
      }

      // [char[i], char[j]] = [char[j], char[i]];

      char[i] = char[j];
      char[j] = char[i];
    }

    return char.join('');
  }
}

class C {
  String reverseVowels(String s) {
    String str = "aeiouAEIOU";
    String dummy = "";
    for (String c in s.split("")) {
      if (str.contains("" + c)) dummy += c;
    }
    List<String> c = s.split("");
    int j = dummy.length - 1;
    for (int i = 0; i < s.length; i++) {
      if (str.contains("" + s[i])) {
        c[i] = dummy[j];
        j--;
      }
    }

    return c.join();
  }
}

class D {
  String reverseVowels(String s) {
    // Create A Pointer Which Is Start From End Of The Array.
    int pointer = s.length - 1;
    // Convert String To Array That Will More Easy To Solve This Problem.
    List<String> ansStr = s.split("");
    int i = 0; // Create A I Veritable That Start From Array Index Of 0
    // Create a All Vowels That Will Help To Check The Value Is Vowel Or Not.
    List<String> vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'];
    // You can use regex like: /[a|e|i|o|u]/gi
    while (i < pointer) {
      // Check Condition That I Should Be Less Then Pointer This Condition Divide The Time Complexity To O(N) To O(LogN).

      if (vowels.contains(ansStr[i]) && vowels.contains(ansStr[pointer])) {
        // Here we check the condition that our I position value and Pointer position value should be vowel
        String tmp = ansStr[i];
        // If condition is true then swap this position value respectively you can create a extra swap function.
        ansStr[i++] = ansStr[pointer];
        ansStr[pointer--] = tmp;
      } else {
        if (vowels.contains(ansStr[i])) {
          // If the Ith position value is vowel then we decrement the Pointer Position value to -1;
          pointer--;
        } else {
          i++; // If the Ith position value does not a vowel then increment the Ith position by +1
        }
      }
    }
    return ansStr.join(""); // Here we convert the array to string
  }
}
